### DISCUSSION TITLE
How to properly turbo eval hooks for externally installed packages

### DISCUSSION THREAD

**Discussion by paw-lu** [zdharma-continuum/zinit#123] [Q&A] [answered]
I have some tools that have either completions or other integrations that are called via `eval` and some command that generates the code.

For example let's look at [theacodes/nox](https://github.com/theacodes/nox). It's installed via another method (`pipx`). And for completion we must call `eval "$(register-python-argcomplete nox)"`. This command must be called _after_ `autoload -U bashcompinit; bashcompinit;`.

Currently implimented:

```zsh
zinit wait lucid light-mode for \
    if"builtin command -v nox > /dev/null 2>&1" \
        atinit"autoload -U bashcompinit; bashcompinit;" \
            is-snippet $HOME/rc_files/nox.zsh \
    ...
```

Where `nox.zsh` simply contains `eval "$(register-python-argcomplete nox)"`. This works as expected.

However this is not efficient, as `register-python-argcomplete` must be run each time to generate the code.

There's a really good [direnv example](https://zdharma-continuum.github.io/zinit/wiki/Direnv-explanation/) that's really similar to my problem, _but_ assumes you're using zinit to compile.

Trying to sloppily convert this, I try:

```zsh
zinit wait lucid light-mode for \
    if"builtin command -v nox > /dev/null 2>&1" \
        atclone"register-python-argcomplete nox > zhook.zsh" \
            atpull'%atclone' \
                src"zhook.zsh" \
                    atinit"autoload -U bashcompinit; bashcompinit;" \
                        is-snippet $HOME/rc_files/nox.zsh \
    ...
```

1. I don't think this works, as it doesn't seem what I put into zhook is actually executing
2. the snippet is actually useleess now, since the actuall command is (meant) to run on `atclone`/`atpull`/`src`. Is there a way to simply run a command in the turbo framework without a snippet or package?

**Comment by paw-lu** [comment]
> Is there a way to simply run a command in the turbo framework without a snippet or package?

It seems this was partially answered by the existence of [zdharma-continuum/null](https://github.com/zdharma-continuum/null), which implies a "package" is always needed here.

**Comment by pschmitt** [comment]
Here's something I usually do for pip(x) packages:

```zsh
zinit wait lucid light-mode as"null" nocompile \
  atclone"pipx install nox; register-python-argcomplete nox > zhook.zsh" \
  atpull"pipx update nox; rm -f zhook.zsh; register-python-argcomplete nox > zhook.zsh" \
  atdelete"pipx uninstall nox" \
  atload"autoload bashcompinit && bashcompinit && source zhook.zsh" \
for theacodes/nox
```

**Reply by paw-lu** [reply]
This is a useful example.

> `for theacodes/nox`

For my understanding, in this case is this line even doing anything?

**Reply by pschmitt** [reply]
Yes, it will fetch the source repo. This allows running the `atpull` commands when there's a new commit in that repo. This obviously only makes sense if you explicitly want to track the main branch though. Here it may result in the `atpull` ice being run too often (not every commit will lead to a new pypi release). If there was a way to make `pipx update nox` return 1 instead of 0 if there is no update then you could at least skip the generation of the `zhook.zsh` file.
It's not perfect, yet ;)

**Comment by pschmitt** [comment, marked as answer]
Oh! There's https://github.com/NICHOLAS85/z-a-eval that's worth a try as well :)

**Reply by paw-lu** [reply]
didn't even know about this!
seems this is a less hacky syntax since you don't have to redirect to a file.

appreciate all the help (and maintainting the fork)!

**Reply by pschmitt** [reply]
🎩
